### ISSUE TITLE
Mocking application class in androidTest instrumented test significantly slows down unrelated tests

### ISSUE THREAD

**Issue by jdvp** [mockk/mockk#956]
### Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

When mocking `Application`, `Activity`, or implementing subclasses in androidTest tests, tests should run as quickly as when mocking other classes. At the very least, other test cases not leveraging MockK should not be impacted.

### Current Behavior

Mocking `Application` or `Activity` significantly slows down test cases.

For example in a suite of 4 tests, adding `mockk<Application>()` in the first case to run significantly slows down each test bringing the run time from roughly 13s to over 3 minutes. Interestingly, the mock was only added to a single test case and only in 1 file but impacts all test cases that run after it regardless of file.

MockKing application class             |  Commenting out the mockk      
:-------------------------:|:-------------------------:
3m11s | 13s
<img src="https://user-images.githubusercontent.com/7111196/198494160-6e524ab6-5fc4-4318-b896-e76ff22249af.png" width="500"/>  |  <img src="https://user-images.githubusercontent.com/7111196/198494164-be9296f8-081e-4a4e-b1fa-5d13108636e1.png" width="500"/> 

#### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Create an androidTest test that does something like clicking a UI element
2. Add a mockk application in a single test case `mockk<Application>()` or can mock Activity: `mockk<Activity>()`. This doesn't even need to be used. Obviously if we really didn't use the mockk, i could just remove it 
4. Run the test and witness test slow down

You can check out my Library Test Example project's branch [`feature/mockk-issue-example`](https://github.com/jdvp/LibraryTestExample/tree/feature/mockk-issue-example) if you want as it's what I used to reproduce since my original sighting was in closed source code. The androidTest items should run slowly locally but if you search for and remove the usage of `mockk<Application>()` in the `ExampleTest` class, it should run much more quickly.

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: `1.13.2`
* OS: tested on `API 31` emulator, compiled with `33`. Saw this in a separate private source project with API 32.
* Kotlin version: `1.7.20`
* JDK version: `11`
* JUnit version: `4.13.2`
* Type of test: android instrumented test

**Comment by jdvp** [comment]
verified this is still an issue with version `1.13.3`

**Comment by kevin-brotcke** [comment]
We experienced this too after converting a test from java to kotlin that mocked activity, took awhile to root cause since we thought the behavior would match mockito. I was able to capture the main thread stack during ANR it causes sometimes:

```
Thread[main,5,main]
  java.lang.Class.getDeclaredMethodsUnchecked(Native Method)
  java.lang.Class.getDeclaredMethods(Class.java:1911)
  io.mockk.proxy.android.advice.Advice$Companion.findMethod(Advice.kt:165)
  io.mockk.proxy.android.advice.Advice$Companion.isOverridden(Advice.kt:159)
  io.mockk.proxy.android.advice.Advice$Companion.access$isOverridden(Advice.kt:155)
  io.mockk.proxy.android.advice.Advice.getOrigin(Advice.kt:38)
  java.lang.reflect.Method.invoke(Native Method)
  io.mockk.proxy.android.AndroidMockKDispatcher.getOrigin(AndroidMockKDispatcher.java:117)
  android.view.ContextThemeWrapper.getResourcesInternal(Unknown Source:14)
  android.view.ContextThemeWrapper.getResources(ContextThemeWrapper.java:128)
  androidx.appcompat.app.AppCompatActivity.getResources(AppCompatActivity.java:609)
  android.content.Context.getDrawable(Context.java:810)
  androidx.core.content.ContextCompat$Api21Impl.getDrawable(ContextCompat.java:1012)
  androidx.core.content.ContextCompat.getDrawable(ContextCompat.java:522)
  androidx.appcompat.widget.ResourceManagerInternal.getDrawable(ResourceManagerInternal.java:149)
  androidx.appcompat.widget.ResourceManagerInternal.getDrawable(ResourceManagerInternal.java:137)
  androidx.appcompat.content.res.AppCompatResources.getDrawable(AppCompatResources.java:66)
  androidx.appcompat.view.menu.MenuItemImpl.getIcon(MenuItemImpl.java:509)
  com.google.android.material.navigation.NavigationBarItemView.initialize(NavigationBarItemView.java:202)
  com.google.android.material.navigation.NavigationBarMenuView.buildMenuView(NavigationBarMenuView.java:735)
  com.google.android.material.navigation.NavigationBarPresenter.updateMenuView(NavigationBarPresenter.java:72)
  com.google.android.material.navigation.NavigationBarView.inflateMenu(NavigationBarView.java:382)
  com.google.android.material.navigation.NavigationBarView.<init>(NavigationBarView.java:276)
  com.google.android.material.bottomnavigation.BottomNavigationView.<init>(BottomNavigationView.java:108)
  com.google.android.material.bottomnavigation.BottomNavigationView.<init>(BottomNavigationView.java:103)
  com.google.android.material.bottomnavigation.BottomNavigationView.<init>(BottomNavigationView.java:98)
  java.lang.reflect.Constructor.newInstance0(Native Method)
  java.lang.reflect.Constructor.newInstance(Constructor.java:343)
  android.view.LayoutInflater.createView(LayoutInflater.java:858)
  android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1010)
  android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:965)
  android.view.LayoutInflater.rInflate(LayoutInflater.java:1127)
  android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1088)
  android.view.LayoutInflater.rInflate(LayoutInflater.java:1130)
  android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1088)
  android.view.LayoutInflater.inflate(LayoutInflater.java:686)
  android.view.LayoutInflater.inflate(LayoutInflater.java:538)
  android.view.LayoutInflater.inflate(LayoutInflater.java:485)
  androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:710)
  androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:195)
  com.example.MyActivity.onCreate(MyActivity.java:100)
```

My red flag here is that mockk methods are being called when this is supposed to be a real activity. It seems like mocking any activity causes side effects to later real activities. Android OS gets stuck in a loop continuously trying to fetch theme, catching error, then trying again. 

Workarounds we found:
* Go back to mockito `Mockito.mock(Activity::class.java)`
* Find a way to avoid mocking activity, use a real activity `Activity()` or create an abstraction layer to mock instead

**Comment by jdvp** [comment]
@kevin-brotcke yep, that is pretty much what we had done too as far as workarounds. A few places where it wasn't really able to be changed we had to ignore for now

Also I have verified still an issue in `1.13.4` unfortunately

**Comment by jdvp** [comment]
Spent a few more minutes looking into it as I ran into a case where I can't just do a workaround. 

I don't have any ideas for the actual test case slowness but I do think that the single slow test shouldn't poison other tests that aren't even using mockk. To me, it appears to be an issue with `unmockkAll` not actually unmockking individual mocks (i.e. those created via `mockk<>()`) and not providing a way to clear these mocks up at the end of tests (related to https://github.com/mockk/mockk/issues/416) 

I found that in my clear method I could do 
```
@After
fun tearDown() {
    //Generally would use with MockK to prevent static/object pollution across tests
    unmockkAll()
    //unmocks all regular Mockks (ideally should be done by [unmockkAll] directly??)
    JvmMockKGateway.defaultImplementationBuilder()
        .stubRepo.allStubs.forEach(Disposable::dispose)
}
```

This helps to ensure that only the test that is mockking `Application` is slowed down (which I would obviously also like to solve):


<img src="https://user-images.githubusercontent.com/7111196/230569217-bf3a8e35-08ee-448b-ba44-95b829e8d627.png" width="500"/> 

I'm also wondering if it's intentional for some reason that dispose isn't just called on all the items in the stubRepository? Still a lot for me to learn about the internal workings of MockK

**Comment by Raibaz** [comment]
This is a very interesting finding, thanks for taking the time to investigate this!

I agree `unmockkAll` should also dispose of existing stubs, it really seems like a bug to me that it doesn't.

Would you be able to send out a PR to add the disposal behavior to `unmockkAll`?

**Comment by jdvp** [comment]
Sorry been having a pretty busy summer so haven't had time to jump in and figure out how to best implement this (since calling 
```
JvmMockKGateway.defaultImplementationBuilder().stubRepo.allStubs.forEach(Disposable::dispose)
```
directly in unmockkAll doesn't appear to be correct haha. Once I get a chance to dig into the lib and really understand how it works (sorry it's beyond me at the moment), I will try to raise an MR (assuming no one else fixes that first!).

**Comment by sylwek845** [comment]
Getting similar issue, but I'm running leak canary for ui test and it causes leaks for all the tests since I added mockk.

**Comment by LitterSun** [comment]
I also encountered the same problem. How did you solve it?
